tests/e2e/test_snapshot.go

```diff
@@ -0,0 +1,194 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/format"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type clusterSnapshot struct {
+	resources []metav1.PartialObjectMetadata
+}
+
+func (s *clusterSnapshot) Resources() []metav1.PartialObjectMetadata {
+	return s.resources
+}
+
+// ResourcesAsComparisonType returns a slice of a helper type that makes comparisons easier
+func (s *clusterSnapshot) ResourcesAsComparisonType() []localComparisonPartialObjectMetadataType {
+	var localList []localComparisonPartialObjectMetadataType
+	for _, r := range s.resources {
+		localList = append(localList, localComparisonPartialObjectMetadataType{
+			GVK:       r.GroupVersionKind(),
+			Namespace: r.GetNamespace(),
+			Name:      r.GetName(),
+		})
+	}
+	return localList
+}
+
+// localComparisonPartialObjectMetadataType holds a version of the minimal information required
+// to compare k8s.io/apimachinery/pkg/apis/meta/v1.PartialObjectMetadata instances
+type localComparisonPartialObjectMetadataType struct {
+	GVK       schema.GroupVersionKind
+	Namespace string
+	Name      string
+}
+
+// snapshotCluster takes a clusterSnapshot of a K8s cluster and
+// stores it into the snapshotCluster instance referenced as input
+func snapshotCluster(snapshottedInfo *clusterSnapshot) bool {
+	return When("Get cluster resources state", Ordered, func() {
+		var kubectlOptions k8s.KubectlOptions
+		var err error
+
+		BeforeAll(func() {
+			By("Acquiring K8s config and context")
+			kubectlOptions, err = kubectlOptionsForCurrentContext()
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		var clusterResourceNames []string
+		var namespacedResourceNames []string
+
+		When("Get api-resources names", func() {
+			It("Get cluster-scoped api-resources names", func() {
+				clusterResourceNames, err = listK8sResourceKinds(kubectlOptions, "", "--namespaced=false")
+				Expect(err).NotTo(HaveOccurred())
+				clusterResourceNames = pruneUnnecessaryClusterResourceNames(clusterResourceNames)
+			})
+			It("Get namespaced api-resources names", func() {
+				namespacedResourceNames, err = listK8sResourceKinds(kubectlOptions, "", "--namespaced=true")
+				Expect(err).NotTo(HaveOccurred())
+				namespacedResourceNames = pruneUnnecessaryNamespacedResourceNames(namespacedResourceNames)
```

Comment:
Same as above, we might want to do a `Expect(namespacedResourceNames).NotTo(BeNil())` after this line, if needed

Reply:
The same comment here https://github.com/banzaicloud/koperator/pull/1018#discussion_r1269389676  applies with the modification that `kubectl api-resources ... ... --namespaced=true` .

Reply:
Done in e791dd6cdbeef047f00f5801f280b2fcb8cafc1b.